### PR TITLE
[CHERI-RISC-V] Fix unwinding signal frames

### PR DESCRIPTION
**Pull request by arichardson**
Three things were breaking signal frame unwinding:
The sigcode assembly is different for purecap (cincoffset instead of
addi), so the current undwinder pattern does not match. Additionally,
this commit updates riscv_fbsd_sigframe_init() to handle the capability
register offsets. Finally, the stack pointer was wrong since
`get_frame_register_unsigned (this_frame, RISCV_SP_REGNUM);` resulted in
a value of 0x43539d70, whereas
`value_as_address(get_frame_register_value(this_frame, RISCV_CSP_REGNUM))`
gave the correct SP value of 0x4353a5c0. Fix this by using `get_frame_sp`
like the other sigframe_init functions. It appears that `get_frame_sp`
correctly reads CSP instead of the (stale?) sp value.

Fixes https://github.com/CTSRD-CHERI/gdb/issues/40

### CODE REVIEW

File: gdb/riscv-fbsd-tdep.c [hunk 1]
```diff
@@ -202,27 +203,36 @@ riscv_fbsd_sigframe_init (const struct tramp_frame *self,
...
+    }
   CORE_ADDR mcontext_addr
-    = (sp
-       + RISCV_SIGFRAME_UCONTEXT_OFFSET
-       + RISCV_UCONTEXT_MCONTEXT_OFFSET);
+    = (frame_sp + ucontext_offset + RISCV_UCONTEXT_MCONTEXT_OFFSET);
+  trad_frame_set_reg_regmap (this_cache, gp_regmap, mcontext_addr,
+			     gp_regmap_size);
+  CORE_ADDR fpregs_addr = mcontext_addr + gp_regmap_size;
+  CORE_ADDR fp_flags_addr = fpregs_addr + RISCV_FBSD_SIZEOF_FPREGSET;
```
Comment (jrtc27) [inline gdb/riscv-fbsd-tdep.c:227]: Leave these where they were and don't reformat fp_flags_addr, you didn't change it

File: gdb/riscv-fbsd-tdep.c [hunk 2]
```diff
@@ -188,7 +188,8 @@ riscv_fbsd_iterate_over_regset_sections (struct gdbarch *gdbarch,
    by the floating point register set.  The floating point register
    set is only valid if the _MC_FP_VALID flag is set in mc_flags.  */
 
-#define RISCV_SIGFRAME_UCONTEXT_OFFSET 		80
+#define RISCV_SIGFRAME_UCONTEXT_OFFSET_RV64	80
+#define RISCV_SIGFRAME_UCONTEXT_OFFSET_PURECAP	112
```
Comment (jrtc27) [inline gdb/riscv-fbsd-tdep.c:192]: LP64 and L64PC128 (or C128 to match the name of the generic SVR4 function)

**Comment by jrtc27** [comment]
Weird, I could have sworn I've unwinded through signal frames on CHERI-RISC-V...

File: gdb/riscv-fbsd-tdep.c (hunk 1 above)
Comment (arichardson) [reply gdb/riscv-fbsd-tdep.c:227]: looks like I accidentally clang-formatted some lines I didn't change.

**Review by bsdjhb — COMMENTED** [review]
Huh, I'm really surprised I hadn't done this already.  It tends to be one of the first things I test (I have little sigframe test program just to test signal frames).  For Morello and MIPS I just used a separate function with separate constants along with the separate tramp_frame rather than adding conditional logic in a shared frame_init function.  (See aarch64-fbsd-tdep.c in the morello-8.3 branch)  I might have a slight preference for keeping the unwinders distinct that way.

**Comment by arichardson** [comment]
> Huh, I'm really surprised I hadn't done this already. It tends to be one of the first things I test (I have little sigframe test program just to test signal frames). For Morello and MIPS I just used a separate function with separate constants along with the separate tramp_frame rather than adding conditional logic in a shared frame_init function. (See aarch64-fbsd-tdep.c in the morello-8.3 branch) I might have a slight preference for keeping the unwinders distinct that way.

I've used separate functions in this version, but kept a shared helper function.

**Review by bsdjhb — APPROVED** [review]

**Pull request merged**